### PR TITLE
feat: upgrading typeorm

### PR DESCRIPTION
**Pull request by YazeedLoonat**
# Pull Request Template

## Issue Overview

This PR addresses [#issue](https://github.com/bloom-housing/bloom/issues/3184)

- [x] This change addresses the issue in full

## Description
This PR upgrades TypeORM to a much newer version

## How Can This Be Tested/Reviewed?
To be honest every part of the system needs to be tested:
**listings**
creating, editing, viewing (public and partners)
**applications**
creating, editing, viewing (public and partners
**users**
creating, editing, viewing
**duplicate listings**
resolving, marking as duplicate, viewing

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [x] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [x] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

File: backend/core/src/migration/1676354438563-typeorm-upgrade.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { MigrationInterface, QueryRunner } from "typeorm";
+
+export class typeormUpgrade1676354438563 implements MigrationInterface {
+    name = 'typeormUpgrade1676354438563'
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`ALTER TABLE "listing_multiselect_questions" DROP CONSTRAINT "FK_92adcb35f2f14e316b4cb12a84e"`);
+        await queryRunner.query(`ALTER TABLE "listing_multiselect_questions" ALTER COLUMN "multiselect_question_id" SET DEFAULT uuid_generate_v4()`);
+        await queryRunner.query(`ALTER TABLE "listings" ALTER COLUMN "afs_last_run_at" SET DEFAULT '1970-01-01'`);
+        await queryRunner.query(`ALTER TABLE "listings" ALTER COLUMN "last_application_update_at" SET DEFAULT '1970-01-01'`);
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend/core/src/migration/1676354438563-typeorm-upgrade.ts:10]: I think it should although I think the migration reset would have addressed it for that PR

File: backend/core/src/seeder/seed.ts [hunk 2]
```diff
@@ -113,7 +113,7 @@ export async function createLeasingAgents(
   )
   await Promise.all([
     leasingAgents.map(async (agent: User) => {
-      const roles: UserRoles = { user: agent, isPartner: true }
+      const roles: UserRoles = { user: agent, isPartner: true, userId: undefined }
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend/core/src/seeder/seed.ts:116]: nice catch! so TypeORM changed how relationships are defined, and now there must be an explicit column (that should match up with the foreign key ) 

if you check the entity definition you should see the update represented there. From what I understood this makes piecing things together by the relationship a little be easier for TypeORM

File: backend/core/package.json [hunk 3]
```diff
@@ -41,12 +41,12 @@
     "@anchan828/nest-sendgrid": "^0.3.25",
     "@google-cloud/translate": "^6.2.6",
     "@nestjs/cli": "^8.2.1",
-    "@nestjs/common": "^8.3.1",
+    "@nestjs/common": "9.3.7",
```
[the remark this answers is not in the document]
Comment (YazeedLoonat) [reply backend/core/package.json:44]: good catch will update!

File: backend/core/src/auth/helpers/user-helpers.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  return getQb(repo).where("user.id = :id", { id }).getOne()
+}
+
+export const findByConfirmationToken = async (repo: Repository<User>, token: string) => {
+  return getQb(repo).where("user.confirmationToken = :token", { token }).getOne()
+}
+
+export const findByResetToken = async (repo: Repository<User>, token: string) => {
+  return getQb(repo).where("user.resetToken = :token", { token }).getOne()
+}
```
Comment (seanmalbert) [inline backend/core/src/auth/helpers/user-helpers.ts:33]: Question (blocking): Do these need to live outside of the user service? For the use cases outside of the user service (like the authz service), isn't it nest js best practice to inject the user service?

File: backend/core/src/auth/services/authz.service.ts [hunk 5]
```diff
@@ -38,7 +39,7 @@ export class AuthzService {
       e = await this.addUserPermissions(e, user)
 
       if (type === "user" && obj?.id && !obj?.jurisdictionId) {
-        const accessedUser = await this.userRepository.findById(obj.id)
+        const accessedUser = await findById(this.userRepository, obj.id)
```
Comment (seanmalbert) [inline backend/core/src/auth/services/authz.service.ts:42]: Suggestion (blocking): If you move `findById` to the user service and inject that (which should already inject the user repo), then this should become simplified and you won't have to pass around the user repo.

File: backend/core/src/auth/services/user.service.ts [hunk 6]
```diff
@@ -44,22 +43,28 @@ import { GetMfaInfoDto } from "../dto/get-mfa-info.dto"
...
 import { Request as ExpressRequest } from "express"
 import { UserProfileUpdateDto } from "../dto/user-profile.dto"
-import { ListingRepository } from "../../listings/db/listing.repository"
+import {
+  getQb,
+  findByEmail,
+  findById,
+  findByConfirmationToken,
+  findByResetToken,
+} from "../helpers/user-helpers"
```
Comment (seanmalbert) [inline backend/core/src/auth/services/user.service.ts:56]: Suggestion (blocking): move these into the user service.

File: backend/core/src/listings/db/listing-helpers.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+  const listing = await createQueryBuilder(listingRepository, "listings")
+    .where(`listings.id = :listingId`, { listingId })
+    .leftJoin("listings.jurisdiction", "jurisdiction")
+    .select(["listings.id", "jurisdiction.id"])
+    .getOne()
+
+  return listing.jurisdiction.id
+}
```
Comment (seanmalbert) [inline backend/core/src/listings/db/listing-helpers.ts:27]: Suggestion (blocking): similar to the user service comments, recommend these get moved into the listing service.

**Review by seanmalbert — CHANGES_REQUESTED** [review]

File: backend/core/package.json [hunk 8]
```diff
@@ -41,12 +41,12 @@
...
-    "@nestjs/core": "^8.3.1",
+    "@nestjs/core": "9.3.7",
     "@nestjs/jwt": "^8.0.0",
-    "@nestjs/passport": "^8.2.1",
-    "@nestjs/platform-express": "^8.3.1",
+    "@nestjs/passport": "9.0.3",
+    "@nestjs/platform-express": "9.3.9",
     "@nestjs/schedule": "^2.1.0",
     "@nestjs/swagger": "5.2.0",
     "@nestjs/throttler": "^2.0.0",
```
Comment (ludtkemorgan) [inline backend/core/package.json:52]: Github won't let me comment on the correct line, but shouldn't this PR be upgrading `typeorm` on line 85?

**Comment by emilyjablonski** [comment]
@YazeedLoonat Any remaining concerns post-bash yesterday? Are you looking for more partners bashing on the preview link?

**Comment by ludtkemorgan** [comment]
@YazeedLoonat I got the backend tests to work. However, I just noticed that there is a skipped test that is failing for the same reason that the Assets one was. And the solution I did for the Assets won't work for the afs tests. Still digging into that now

**Comment by emilyjablonski** [comment]
I think there are still bugs from bash to resolve (maybe? @YazeedLoonat)

**Comment by emilyjablonski** [comment]
@YazeedLoonat Just confirming, I think this one is not ready yet, is that right?

**Comment by YazeedLoonat** [comment]
@emilyjablonski @seanmalbert @ColinBuyck @ludtkemorgan 
Hey everyone, made updates across a lot of this PR if you wanna give it another go
🤞 everything should be working on this one!

**Comment by emilyjablonski** [comment]
@YazeedLoonat Looks like it's not building as a heads up!

**Comment by YazeedLoonat** [comment]
Update on this one:

Currently this will need to be put on pause. 
We ran into an issue where the JurisdicitionResolverService used in the Email Service is not available (either by the automatic nest js instantiation step, or using ModuleRef.resolve)

The most likely cause of this is a circular dependency issue that we have throughout the application.

From looking at NestJS documentation particularly (https://docs.nestjs.com/fundamentals/circular-dependency) we have both circular dependencies and leverage a per request scope. 
https://github.com/nestjs/nest/issues/5778

This combination is something Nest does not support, and we would need to refactor our backend to either get rid of the per request scope, or the leveraging of circular dependencies. 

Either rewrite is likely a large effort and so a spike ticket has been written up to look into these options.

**Comment by seanmalbert** [comment]
> Update on this one:
> 
> Currently this will need to be put on pause. We ran into an issue where the JurisdicitionResolverService used in the Email Service is not available (either by the automatic nest js instantiation step, or using ModuleRef.resolve)
> 
> The most likely cause of this is a circular dependency issue that we have throughout the application.
> 
> From looking at NestJS documentation particularly (https://docs.nestjs.com/fundamentals/circular-dependency) we have both circular dependencies and leverage a per request scope. [nestjs/nest#5778](https://github.com/nestjs/nest/issues/5778)
> 
> This combination is something Nest does not support, and we would need to refactor our backend to either get rid of the per request scope, or the leveraging of circular dependencies.
> 
> Either rewrite is likely a large effort and so a spike ticket has been written up to look into these options.
> 
> [nestjs/nest#5487](https://github.com/nestjs/nest/issues/5487) is a similar issue to what we are running into, strict mode was not the solution

@YazeedLoonat , I did not run into this issue. I did run into a small issue in the jurisdiction resolver service and fixed it. Locally, I'm able to login with MFA with both email and sms.

File: backend/core/test/afs/afs.e2e-spec.ts [hunk 9]
```diff
@@ -31,7 +30,7 @@ import cookieParser from "cookie-parser"
 declare const expect: jest.Expect
 jest.setTimeout(30000)
 
-describe("ApplicationFlaggedSets", () => {
+describe.skip("ApplicationFlaggedSets", () => {
```
Comment (ludtkemorgan) [inline backend/core/test/afs/afs.e2e-spec.ts:33]: We should probably figure out how to resolve the issues with this test rather than skipping it

Comment (ludtkemorgan) [reply backend/core/test/afs/afs.e2e-spec.ts:33]: Taking a quick pass over it, it looks like there are two issues. One is an easy fix that will solve the logger issue. The other one might be a bit tougher that is either a fix to the test or there is potentially an actual bug in the cron job code

**Review by ludtkemorgan — CHANGES_REQUESTED** [review]
I haven't found any functional issues, but still investigating how to solve that broken skipped test. There is a chance the AFS code has an issue, but more likely we just need to update the test somehow

**Comment by YazeedLoonat** [comment]
@ludtkemorgan @seanmalbert @emilyjablonski @ColinBuyck this should be ready for re-review now!

**Review by ludtkemorgan — APPROVED** [review]
🦖 🌮

**Comment by ColinBuyck** [comment]
@YazeedLoonat do you mind resolving the merge conflicts when you get a chance? Hoping to test this with listings export!

**Comment by YazeedLoonat** [comment]
hey @ColinBuyck merge conflicts have been resolved, lemme know what you think

**Review by ColinBuyck — CHANGES_REQUESTED** [review]
I'm unfortunately getting a 500 when attempting to resolve an application set
Screenshot 2023-05-15 at 2 16 40 PM

**Comment by YazeedLoonat** [comment]
Hey @ColinBuyck this should be good to go again

**Review by ColinBuyck — APPROVED** [review]
LGTM 🍌  I rate this 12/10!

**Pull request merged**